### ISSUE TITLE
Tracking issue: Introduce QGM for SQL-to-SQL transforms

### ISSUE THREAD

**Issue by uce** [MaterializeInc/materialize#8073]
## Summary

We want to introduce a new stage in the query compilation pipeline between HIR and MIR that is based on the [Query Graph Model](http://projectsweb.cs.washington.edu/research/projects/db/weld/pirahesh-starburst-92.pdf). The main goal of this stage is to allow SQL-specific transformations that are currently hard to express in later stages of the optimization pipeline.

~~We want to replace the `HirRelationExpr` stage of the optimization pipeline with a new representation based on the [Query Graph Model](http://projectsweb.cs.washington.edu/research/projects/db/weld/pirahesh-starburst-92.pdf) representation.~~

~~The goal of this refactoring is to alleviate pain points of the current representation (e.g. complexity of name resolution, functional dependency analysis, simplify lowering) and introduce a normalization stage for SQL-specific transformations that are currently hard to express in later stages of the optimization pipeline.~~

## Goals

* Feature is enabled/disabled with a CLI feature flag
* Support all HIR plans
* Support lowering QGM plans into MIR

## Non-Goals

* Resolve pain points of HIR (such as name resolution or full support for lateral joins).
* Originally we planned to replace HIR with QGM, but this requires a lot of work before we can actually get to improving plans and switching the feature on by default. This can be an optional follow-up.

## Design

The following design documents track the evolution of the design implemented in this meta-issue.

1. [QGM as high level representation for SQL queries (2021/07/07)](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20210707_qgm_sql_high_level_representation.md)

## Implementation

The implementation roadmap consists of different areas, which in general are presented below in dependency order, although certain parts can be tackled in parallel.

The workflow is as follows:
- Issues that need grooming are marked with *needs-discussion* or *needs-details* labels.
- Issues without those labels can be picked up. When you pick up an issue, please
  - assign yourself to that issue, and
  - add it to [the "In Progress" column of the "Optimization and Dataflow" project](https://github.com/orgs/MaterializeInc/projects/16).
- Follow-up items should be first added as an item to one of the following sections. @aalexandrov will periodically:
  - convert new items to issues labeled with *needs-discussion* and/or *needs-details*, and
  - organize a grooming session to nail down the details of new issues.
   

### Support `SELECT [<literal>]*` with QGM
- [x] #8801
  - Query Graph Model definition
  - Skeleton for Hir ⇒ QGM translation 
  - Add support for unit testing and debug mode for QGM plans
- [x] #8969
- [x] #8863
- [x] #8958
- [x] #9128
- [x] #9125
- [x] #9127

### Support scalar expressions
- [x] #9140
- [x] #9141
- [x] #9142

### Support HIR ⇒ QGM ⇒ MIR for all variants

Edited 3/25/22: instead of going from HIR ⇒ QGM ⇒ MIR right away, we are trying HIR ⇒ QGM ⇒ HIR in order to avoid figuring out decorrelating QGM for the time being. Currently as of #11243, there is HIR ⇒ QGM ⇒ HIR support for:
* outer join
* select boxes (with distinct, without limit/order by/ offset)
* get
* values
* lets
No support for any correlated queries yet, but a prototype can be found [here](https://github.com/wangandi/materialize/commit/743df072e4eeb24398b10c05e039aee711e8590f). 

- [x] #9180
- [x] #9138
- [x] #9139
- [x] #9129
- [x] #9130
- [x] #9131
- [x] #9136
- [x] #9132
- [x] #9133
- [x] #9134
- [x] #11227
- [ ] #9632
- [x] #9137

### Add `Except` and `Intersect` operators to Hir
- [x] #9145
- [x] #9148

### Support correlated subqueries
Needs scoping

### Constraint validation
- [x] #11081
- [ ] Stuff in https://github.com/MaterializeInc/materialize/issues/8958#issuecomment-1007361231?

### Testing
- [x] #10518
- [x] #9347

### Derived Attributes
- [x] #10237
- [x] #10238
- [x] #10409
- [x] #9601
- [x] #10254

### Rewrites
- [x] #9143
- [x] #10213
- [x] #9144

### Final Steps
- [x] #8993

### Technical Debt
- [x] #10223
- [x] #10224

## Open Questions

* How do we consolidate the different optimization phases into a single coherent API (see [discussion in #8990](https://github.com/MaterializeInc/materialize/pull/8990#pullrequestreview-808586638))?

## Related Issues

The following issues should be unblocked once this ticket has been done:
- #6932
- #8574
- #8395
- #8391
- #8380
- #8223
- #7707
- #6988
- #6490
- #8200
- https://github.com/MaterializeInc/product/issues/125

**Comment by leiysky** [comment]
Have you considered using a Volcano/Cascades style approach?
